src/pymor/algorithms/rand_la.py

```diff
@@ -2,21 +2,266 @@
+    A
+        The |Operator| whose range is to be found.
+    range_product
+        Inner product |Operator| of the range of `A`. Determines the basis orthogonalization and the
+        error norm.
+    source_product
+        Inner product |Operator| of the source of `A`. Determines the basis orthogonalization (only
+        if `subspace_iterations` is greater than zero) and the error norm.
+    subspace_iterations
+        The number of subspace iterations (defaults to zero).
+        This can be used to increase the precision in the cases where the spectrum of `A` does not
+        decay rapidly (at the expense of two additional passes over `A` per subspace iteration).
+    lambda_min
+        A lower bound for the smallest eigenvalue of `source_product`. Defaults to `None`.
+        If `None` and a `source_product` is given, the smallest eigenvalue will be computed
+        using :func:`randomized GHEP <pymor.algorithms.rand_la.randomized_ghep>`.
+    complex
+        If `True`, complex valued random vectors will be chosen.
+    self_adjoint
+        If `True`, the calculation of the adjoint will be skipped and `op` will be used as its
+        adjoint when calculating the subspace iterations. Defaults to `False`
+    """
+
+    def __init__(self, A, range_product=None, source_product=None, subspace_iterations=0, lambda_min=None,
+                 complex=False, self_adjoint=False):
+        assert isinstance(A, Operator)
+        assert 0 <= subspace_iterations and isinstance(subspace_iterations, int)
+        if range_product is None:
+            range_product = IdentityOperator(A.range)
+
+        if source_product is None:
+            source_product = IdentityOperator(A.source)
+        assert isinstance(range_product, Operator)
+        assert source_product.source == source_product.range == A.source
+        assert isinstance(source_product, Operator)
+        assert source_product.source == source_product.range == A.source
+        assert lambda_min is None or isinstance(lambda_min, Number)
+        assert isinstance(complex, bool)
+        assert isinstance(self_adjoint, bool)
+
+        self.__auto_init(locals())
+        self._Q = [self.A.range.empty()]
+        for _ in range(subspace_iterations):
+            self._Q.append(self.A.source.empty())
+            self._Q.append(self.A.range.empty())
+        self._Q = tuple(self._Q)
+        self._samplevecs = self.A.range.empty()
+        self._adjoint_op = A if self_adjoint else AdjointOperator(A, range_product=range_product,
+                                                                  source_product=source_product)
+
+    @cached
+    def _lambda_min(self):
+        if isinstance(self.source_product, IdentityOperator):
+            return 1
+        elif self.lambda_min is None:
+            with self.logger.block('Estimating minimum singular value of source_product ...'):
+                return randomized_ghep(InverseOperator(self.source_product), n=1)[0]
```

Comment:
Have you actually tested if this works well?

Reply:
I did some tests with random matrices up to order 1000 and it seems to work fine.

Reply:
Actually, it does not...

Reply:
Forgot to invert the eigenvalue. Now it looks alright if the smallest eigenvalue is not smaller than machine epsilon.